File: runme/mcgen/local/datan/exasrc/e2reg.f
      PROGRAM E2REG
      IMPLICIT DOUBLE PRECISION (A-H,O-Z)
      PARAMETER(N=10,NR=10,NPL=200)
      DIMENSION T(N),Y(N),DELTAY(N),X(NR),CHI2(NR)
      DIMENSION B(NR,NR),A(N,NR)
      DIMENSION XPL(NPL,NR),YPL(NPL,NR),NCOL(NR)
      DIMENSION DATSX(N),DATCOV(N)
      CHARACTER*75 TX,TY,CAPT
      PARAMETER (ZERO=0.D0,ONE=1.D0)
      DATA T
     +/-.9D0,-.7D0,-.5D0,-.3D0,-.1D0,.1D0,.3D0,.5D0,.7D0,.9D0/
      DATA Y /81.D0,50.D0,35.D0,27.D0,26.D0,
     +60.D0,106.D0,189.D0,318.D0,520.D0/
C identify program to user
      WRITE(*,*)' Program E2REG performs polynomial regression'
      WRITE(*,*)' and presents data and polynomials graphically'
      WRITE(*,*)' '
C set errors
      DO 10 I=1,N
        DELTAY(I)=SQRT(Y(I))
10    CONTINUE
C perform polynomial regression
      CALL REGPOL(T,Y,DELTAY,N,NR,X,B,A,CHI2)
C output results in alphanumeric form
      WRITE(*,*)' T '
      WRITE(*,'(10F7.2)')T
      WRITE(*,*)' Y'
      WRITE(*,'(10F7.2)')Y
      WRITE(*,*)' DELTAY'
      WRITE(*,'(10F7.2)')DELTAY
      WRITE(*,*)' X'
      WRITE(*,'(10F7.2)')X
      WRITE(*,*)' CHI2'
      WRITE(*,'(10F7.2)')CHI2
C ask for number of workstation
      CALL GRNBWS()
      WRITE(*,*)' Please, enter number of workstation:'
      WRITE(*,*)' >'
      READ(*,*)NWS
      WRITE(*,*)' Enter maximum number of terms in polynomials'
      WRITE(*,*)' to be plotted (>0, <11)'
      WRITE(*,*)' >'
      READ(*,*)NPOL
C set errors along abscissa and covariances to zero for graphics
      DO 15 I=1,N
        DATSX(I)=0.D0
        DATCOV(I)=0.D0
15    CONTINUE
C compute polylines for polynomials of degrees 0,1,2,....
      T0=-1.5D0
      DT=3.D0/DBLE(NPL-1)
      DO 40 I=1,NPL
        DO 30 J=1,NPOL
          XPL(I,J)=T0+(I-1)*DT
          YPL(I,J)=ZERO
          DO 27 L=1,J
            D=B(L,1)
            IF(L.GT.1) THEN
              DO 25 K=2,L
                D=D+B(L,K)*XPL(I,J)**(K-1)
25            CONTINUE
            END IF
            YPL(I,J)=YPL(I,J)+X(L)*D
27        CONTINUE
30      CONTINUE
40    CONTINUE
      DO 50 J=1,10
        NCOL(J)=4
50    CONTINUE
C prepare texts for graphics
      TX='t'
      LTX=1
      TY='y'
      LTY=1
      CAPT=' '
      LCAPT=1
      NMARK=5
      SCALEF=.5D0
C call graphics routine for data points and multiple polylines
      CALL GRDTMC(XPL,YPL,NPL,NPOL,NCOL,NMARK,SCALEF,T,Y,DATSX,
     +DELTAY,DATCOV,N,TX,LTX,TY,LTY,CAPT,LCAPT,NWS)
      END

File: runme/mcgen/local/datan/exasrc/e2reg.c
#include <stdio.h>
#include <math.h>
#include <string.h>
#include "datsrc.h"
#include "grsrc.h"

/* Table of constant values */

static integer c__10 = 10, c__200 = 200;

struct {
    double xpl[2000] /* was [200][10] */, ypl[2000] /* was [200][10] */;
} LIBDATA e1rcom;

main(void)
{
    /* Initialized data */ /* Local variables */
    static integer i, j, k, l, ncol[10], npol, lcapt, nmark, ltx, lty, nws;
    static double t[10] = { -.9,-.7,-.5,-.3,-.1,.1,.3,.5,.7,.9 },
                  y[10] = { 81.,50.,35.,27.,26.,60.,106.,189.,318.,520. },
                  a[100] /* was [10][10] */, b[100] /* was [10][10] */,
                  d, x[10], datsx[10], t0, dt, scalef, datcov[10],
                  deltay[10], chi2[10];
    static char capt[75], tx[75], ty[75];
    /* System generated locals */
    integer i__1, i__2, i__3, i__4;
/* identify program to user */
    printf("%s\n", "Program E2REG performs polynomial regression");
    printf("%s\n\n", "and presents data and polynomials graphically.");
/* set errors */
    for (i = 1; i <= 10; ++i) {
        deltay[i - 1] = sqrt(y[i - 1]);
    }
/* perform polynomial regression */
    regpol_(t, y, deltay, &c__10, &c__10, x, b, a, chi2);
/* output results in alphanumeric form */
    printf("%s\n", "T");
    for (i = 1; i <= 10; ++i)
        printf("%7.2f", t[i-1]);
    printf("\n");
    printf("%s\n", "Y");
    for (i = 1; i <= 10; ++i)
        printf("%7.2f", y[i-1]);
    printf("\n");
    printf("%s\n", "DELTAY");
    for (i = 1; i <= 10; ++i)
        printf("%7.2f", deltay[i-1]);
    printf("\n");
    printf("%s\n", "X");
    for (i = 1; i <= 10; ++i)
        printf("%7.2f", x[i-1]);
    printf("\n");
    printf("%s\n", "CHI2");
    for (i = 1; i <= 10; ++i)
        printf("%7.2f", chi2[i-1]);
    printf("\n");
/* ask for number of workstation */
    grnbws_();
    printf("%s\n", "Please, enter number of workstation:");
    printf("%s", "> ");
    scanf("%li", &nws);
    printf("%s\n", "Enter maximum number of terms in polynomials");
    printf("%s\n", "to be plotted (>0, <11):");
    printf("%s", "> ");
    scanf("%li", &npol);
    (void)getchar();
/* set errors along abscissa and covariances to zero for graphics */
    for (i = 1; i <= 10; ++i) {
        datsx[i - 1] = 0.;
        datcov[i - 1] = 0.;
    }
/* compute polylines for polynomials of degrees 0,1,2,.... */
    t0 = -1.5;
    dt = .01507537688442211;
    for (i = 1; i <= 200; ++i) {
        i__1 = npol;
        for (j = 1; j <= i__1; ++j) {
            e1rcom.xpl[i + j * 200 - 201] = t0 + (i - 1) * dt;
            e1rcom.ypl[i + j * 200 - 201] = 0.;
            i__2 = j;
            for (l = 1; l <= i__2; ++l) {
                d = b[l - 1];
                if (l > 1) {
                    i__3 = l;
                    for (k = 2; k <= i__3; ++k) {
                        i__4 = k - 1;
                        d += b[l + k * 10 - 11] *
                             pow(e1rcom.xpl[i + j * 200 - 201], (double)i__4);
                    }
                }
                e1rcom.ypl[i + j * 200 - 201] += x[l - 1] * d;
            }
        }
    }
    for (j = 1; j <= 10; ++j) {
        ncol[j - 1] = 4;
    }
/* prepare texts for graphics */
    memset(tx, ' ', 75);
    memset(ty, ' ', 75);
    memset(capt, ' ', 75);
    tx[0] = 't';
    ty[0] = 'y';
    ltx = 1;
    lty = 1;
    lcapt = 1;
    nmark = 5;
    scalef = .5;
/* call graphics routine for data points and multiple polylines */
    grdtmc_(e1rcom.xpl, e1rcom.ypl, &c__200, &npol, ncol, &nmark, &scalef,
           t, y, datsx, deltay, datcov, &c__10, tx, &ltx, ty, &lty, capt,
           &lcapt, &nws);
    return 0;
} /* main */

